### DISCUSSION TITLE
Dynamic button type and automatic submitting

### DISCUSSION THREAD

**Discussion by crowd-dong** [jaredpalmer/formik#3074] [Q&A]
Greetings formik maintainers and fellow formik user.

I have a question regarding a behavior I am see with my multi step formik form. 

Basically i have a component that maintains a `step` state that will show different content depending on the step.  There is a function that renders a button for the form.  Depending on the state it either renders a `<button type="button" onClick={handleIncrementStep} />` which will only increment to the next step.  Once the user navigate to to the last `step`  the logic for the button will change the type to `<button type="submit" />`

My issue that am seeing is that when the user goes to the last step, where the button changes from `type=button` to `type=submit` formik actually run the `onSubmit` logic.  

I am wondering why this is the case where Formik see a click to `type=button` but treats it as a click to `type=submit`?

Am i doing something wrong and I am not properly building using formik or is there an issue with formik or maybe with react?
Thank you for your time.

here is a codesandbox that can demostrate the behavior.
https://codesandbox.io/s/formik-example-forked-obx2y?file=/index.js

**Comment by johnrom** [comment]
This is definitely a React thing. If you think about how React works, basically it duplicates every single event sent from the elements to the document, and sends them through a React event manager process. So when you do onClick, you're really doing something like this:

```
// ... it's even more complex than this, but this is an oversimplified version
input.onClick
-> parent.onClick
  -> form.onClick
    -> etc
      -> document.onClick
        -> ReactHandlers.where(target => target === input).onClick()
          // parentHandlers.onClick(), etc
            // ReactDocumentRoot.onClick(), onClickCapture()
          // parentHandlers.onClickCapture()
        <- ReactTarget.onClickCaptureHandlers
      <- document.onClickCapture
    <- etc
  <- form.onClickCapture
<- parent.onClickCapture
input.onClickCapture
```

Now, if you think really carefully about what is happening, React.clickHandlers for your button trigger an update to state to update the step, then trigger a re-render. The re-render actually "reconciles" your button to have a new attribute of type="submit" instead of creating a new button. This means by the time your HTML element onclick enters the capture phase, it has already become of type="submit", and the form will trigger a submit. There are three options.

I'd note that the above description is "basically" what is happening, and the internal technical details are probably even more esoteric.

- Simpler but less realistic: Add a `key` to your separate buttons so they are considered separate HTML elements.
- Even simpler but kind of weird, add a setTimeout to setSteps() so it is triggered on the next render.
- More complex, but how React works -- update the clickHandler callback when `steps` changes, using `useCallback`.

All of those options are quick fixes, but there is an underlying issue with your form you haven't discovered yet. When a "page of the form" is submitted, it should usually actually be a form submit. It should be validated, and it should trigger onSubmit. You can use a submit button the whole time, and check if you're on the final page in the onSubmit function. This allows people to hit "enter" on your form to go to the next page, and if you use field-level validations, you can make sure the form is valid before proceeding to the next page using Formik's internal mechanics.

If you use Form level validations (Yup or Manual), it's a bit more complicated as you'll have to choose Formik's validator based on the step.

**Reply by johnrom** [reply]
You can see an example of this here: https://codesandbox.io/s/formik-example-forked-u73di?file=/index.js

I console.log whether you are clicking a type="button" or a type="submit", as well as whether React thinks you're clicking a button or submit. You can uncomment the `key` to see that changing elements works, except that now submit runs into the same issue backwards. When you submit, the click runs, removes the submit button from the DOM on reconcile, and now the click does not know that the submit button belonged to this form, so it doesn't trigger submit. (Adding an ID to your form and the attribute `form={yourId}` might work.)

**Reply by johnrom** [reply]
In all above cases, I'd recommend removing onClick from submit button and doing special logic there, because a user could hit enter.

**Comment by crowd-dong** [comment]
Thank you so much for the quick answer!  I will adjust my code accordingly. @johnrom
